python/ray/actor.py

```diff
@@ -431,7 +433,14 @@ def _ray_from_modified_class(
         # Make sure the actor class we are constructing inherits from the
         # original class so it retains all class properties.
         class DerivedActorClass(cls, modified_class):
-            pass
+            def __init__(self, *args, **kwargs):
+                try:
+                    cls.__init__(self, *args, **kwargs)
+                except Exception as e:
+                    if isinstance(e, ActorClassInheritanceException):
+                        raise e
+
+                    modified_class.__init__(self, *args, **kwargs)
```

Comment:
So I noted that `ActorClass.__init__` will anyway raise a `TypeError` whenever it will be inherited. To exactly figure out whether the exception is due to inheritance of `ActorClass`, I created a new class `ActorClassInheritanceException(TypeError)`. Now, whenever this will be raised, then `DerivedActorClass` will get a clear signal about inheritance of `ActorClass`. In other cases, it will be safe to conclude (AFAICT) that user called `__init__` method of their class and we will proceed normally. IMHO, this is a better and more robust solution which just depends on a simple signal i.e., raising a particular exception in a specific event. It doesn't matter how inheritance is prevented as in the end we just need to raise `ActorClassInheritanceException` and all other code will be able to detect that easily.

@pcmoritz Does this look good?